File: searchlib/src/vespa/searchlib/expression/integerbucketresultnode.cpp
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
#include "integerbucketresultnode.h"
#include <vespa/vespalib/objects/visit.h>

namespace search::expression {

IMPLEMENT_RESULTNODE(IntegerBucketResultNode, BucketResultNode);

IntegerBucketResultNode IntegerBucketResultNode::_nullResult;

size_t
IntegerBucketResultNode::hash() const
{
    return _from;
}

int
IntegerBucketResultNode::onCmp(const Identifiable & b) const
{
    int64_t f1(_from);
    int64_t f2(static_cast<const IntegerBucketResultNode &>(b)._from);
    if (f1 < f2) {
        return -1;
    } else if (f1 > f2) {
        return 1;
    } else {
        int64_t t1(_to);
        int64_t t2(static_cast<const IntegerBucketResultNode &>(b)._to);
        if (t1 < t2) {
            return -1;
        } else if (t1 > t2) {
            return 1;
        }
    }
    return 0;
}

int IntegerBucketResultNode::contains(const IntegerBucketResultNode & b) const
{
    int64_t diff(_from - b._from);
    if (diff < 0) {
        return (_to < b._to) ? -1 : 0;
    } else {
        return (_to > b._to) ? 1 : 0;
    }
}

void
IntegerBucketResultNode::visitMembers(vespalib::ObjectVisitor &visitor) const
{
    visit(visitor, _fromField, _from);
    visit(visitor, _toField, _to);
}

vespalib::Serializer &
IntegerBucketResultNode::onSerialize(vespalib::Serializer & os) const
{
    return os.put(_from).put(_to);
}

vespalib::Deserializer &
IntegerBucketResultNode::onDeserialize(vespalib::Deserializer & is)
{
    return is.get(_from).get(_to);
}

}

// this function was added by ../../forcelink.sh
void forcelink_file_searchlib_expression_integerbucketresultnode() {}

File: searchlib/src/main/java/com/yahoo/searchlib/expression/IntegerBucketResultNode.java
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.searchlib.expression;

import com.yahoo.vespa.objects.Deserializer;
import com.yahoo.vespa.objects.ObjectVisitor;
import com.yahoo.vespa.objects.Serializer;

/**
 * This is an integer bucket value
 *
 * @author <a href="mailto:havardpe@yahoo-inc.com">Haavard Pettersen</a>
 * @author Simon Thoresen Hult
 */
public class IntegerBucketResultNode extends BucketResultNode {

    public static final int classId = registerClass(0x4000 + 101, IntegerBucketResultNode.class);
    private long from = 0; // bucket start, inclusive
    private long to = 0; // bucket end, exclusive

    /**
     * Constructs an empty result node.
     */
    public IntegerBucketResultNode() {
        // empty
    }

    /**
     * Create a bucket with the given limits
     *
     * @param from bucket start
     * @param to   bucket end
     */
    public IntegerBucketResultNode(long from, long to) {
        this.from = from;
        this.to = to;
    }

    /**
     * Obtain the bucket start
     *
     * @return bucket start
     */
    public long getFrom() {
        return from;
    }

    /**
     * Obtain the bucket end
     *
     * @return bucket end
     */
    public long getTo() {
        return to;
    }

    @Override
    public boolean empty() {
        return to == from;
    }

    @Override
    protected int onGetClassId() {
        return classId;
    }

    @Override
    protected void onSerialize(Serializer buf) {
        buf.putLong(null, from);
        buf.putLong(null, to);
    }

    @Override
    protected void onDeserialize(Deserializer buf) {
        from = buf.getLong(null);
        to = buf.getLong(null);
    }

    @Override
    protected int onCmp(ResultNode rhs) {
        if (classId != rhs.getClassId()) {
            return (classId - rhs.getClassId());
        }
        IntegerBucketResultNode b = (IntegerBucketResultNode)rhs;
        if (from < b.from) return -1;
        if (from > b.from) return 1;
        if (to < b.to) return -1;
        if (to > b.to) return 1;
        return 0;
    }

    @Override
    public int hashCode() {
        return super.hashCode() + (int)from + (int)to;
    }

    @Override
    public void visitMembers(ObjectVisitor visitor) {
        super.visitMembers(visitor);
        visitor.visit("from", from);
        visitor.visit("to", to);
    }
}
